app/models/spree/stock/availability_validator_decorator.rb

```diff
@@ -0,0 +1,41 @@
+Spree::Stock::AvailabilityValidator.class_eval do
+  def validate(line_item)
+    # OFN specific check for in-memory :skip_stock_check attribute
+    return if line_item.skip_stock_check
+
+    quantity_to_validate = line_item.quantity - quantity_in_shipment(line_item)
+    return if quantity_to_validate < 1
+
+    validate_quantity(line_item, quantity_to_validate)
+  end
+
+  private
+
+  # This is an adapted version of a fix to the inventory_units not being considered here.
+  # See #3090 for details.
+  # This can be removed after upgrading to Spree 2.4.
+  def quantity_in_shipment(line_item)
+    shipment = line_item_shipment(line_item)
+    return 0 unless shipment
+
+    units = shipment.inventory_units_for(line_item.variant)
+    units.count
+  end
+
+  def line_item_shipment(line_item)
+    return line_item.target_shipment if line_item.target_shipment
+    return line_item.order.shipments.first if line_item.order.present? && line_item.order.shipments.any?
+  end
+
+  # This is the spree v2.0.4 implementation of validate
+  # But using the calculated quantity instead of the line_item.quantity.
+  def validate_quantity(line_item, quantity)
+    quantifier = Spree::Stock::Quantifier.new(line_item.variant_id)
+    return if quantifier.can_supply? quantity
+
+    variant = line_item.variant
+    display_name = %Q{#{variant.name}}
+    display_name += %Q{ (#{variant.options_text})} unless variant.options_text.blank?
+    line_item.errors[:quantity] << Spree.t(:out_of_stock, :scope => :order_populator, :item => display_name.inspect)
```

Comment:
we should stop using hashrockets for hashes and use symbols IMO. It's been 7 years since this syntax was introduced :see_no_evil:

Reply:
yes, I know. but these couple of lines are spree code as is, I'd leave it as it is to make it easier for the next developer to see that.

Reply:
I agree to keep Spree code in it's original form. One improvement would be to add a reference to the original code: https://github.com/spree/spree/blob/v2.0.4/core/app/models/spree/stock/availability_validator.rb

I wouldn't have introduced the guard clause either, but I'm happy with the current version.

Just to play around with ideas. If we wanted to keep and use the original code, we could have wrapped the line item in a new proxy object that acts like the original line item but returns our given quantity. That would make the separation clearer, but it's a lot of hassle for this little change. I'm still happy with the current version. :+1:

Reply:
I am actually fixing this code (at least the hash rocckets!) in the next PR here :-) https://github.com/openfoodfoundation/openfoodnetwork/pull/3232/files#diff-d5194b883a5ff6787c65bd6c4175a4b2R42